activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -442,10 +442,21 @@ def using_limitable_reflections?(reflections)
       def find_with_ids(*ids)
         raise UnknownPrimaryKey.new(@klass) if primary_key.nil?
 
-        expects_array = ids.first.kind_of?(Array)
+        expects_array = if primary_key.is_a?(Array)
```

Comment:
I have mixed feelings about encoding `query_constraints_list` == multi-keyed. If we can think of a better name, perhaps a `primary_key_is_an_array?` helper could defer that question?

Reply:
... or perhaps even `@klass.is_this_a_list_of_ids_rather_than_just_one?(ids.first)`?

Reply:
`@klass.composite_primary_key?` sounds reasonable to me. Do you think it would be more reasonable to have a separate PR to introduce `@klass.composite_primary_key?` abstraction? We have more places that still do `primary_key.is_a?(Array)` and I would prefer changing those all at once while also having a separate PR so we can discuss implementation details like whether we should memoize the value or whether the value should be always derived from `primary_key.is_a?(Array)` vs explicitly setting `@composite_primary_key = true` once when we derive it from schema

Reply:
Extract it as a method here, in a separate commit, replacing existing usage as appropriate, but without introducing new logic / caching -- we can consider that part later, but it's fine to do a straight "extract method" here.